lib-ee/emqx_license/src/emqx_license.erl

```diff
@@ -0,0 +1,147 @@
+    del_license_hook(),
+    emqx_conf:remove_handler(?CONF_KEY_PATH),
+    emqx_license_cli:unload().
+
+-spec update_file(binary() | string()) ->
+    {ok, emqx_config:update_result()} | {error, emqx_config:update_error()}.
+update_file(Filename) when is_binary(Filename); is_list(Filename) ->
+    Result = emqx_conf:update(
+               ?CONF_KEY_PATH,
+               {file, Filename},
+               #{rawconf_with_defaults => true, override_to => cluster}),
+    handle_config_update_result(Result).
+
+-spec update_value(binary() | string()) ->
+    {ok, emqx_config:update_result()} | {error, emqx_config:update_error()}.
+update_value(Value) when is_binary(Value); is_list(Value) ->
+    Result = emqx_conf:update(
+               ?CONF_KEY_PATH,
+               {value, Value},
+               #{rawconf_with_defaults => true, override_to => cluster}),
+    handle_config_update_result(Result).
+
+%%------------------------------------------------------------------------------
+%% emqx_hooks
+%%------------------------------------------------------------------------------
+
+check(_ConnInfo, AckProps) ->
+    #{max_connections := MaxClients} = emqx_license_checker:limits(),
+    case check_max_clients_exceeded(MaxClients) of
+        true ->
+            case MaxClients of
+                0 -> ?LOG(error, "Connection rejected due to the license expiration");
+                _ -> ?LOG(error, "Connection rejected due to max clients limitation")
+            end,
+            {stop, {error, ?RC_QUOTA_EXCEEDED}};
+        false ->
+            {ok, AckProps}
+    end.
+
+%%------------------------------------------------------------------------------
+%% emqx_config_handler callbacks
+%%------------------------------------------------------------------------------
+
+pre_config_update(_, Cmd, Conf) ->
+    {ok, do_update(Cmd, Conf)}.
+
+post_config_update(_Path, _Cmd, NewConf, _Old, _AppEnvs) ->
+    case read_license(NewConf) of
+        {ok, License} ->
+            {ok, emqx_license_checker:update(License)};
+        {error, _} = Error -> Error
+    end.
+
+%%------------------------------------------------------------------------------
+%% Private functions
+%%------------------------------------------------------------------------------
+
+add_license_hook() ->
+    ok = emqx_hooks:put('client.connect', {?MODULE, check, []}).
+
+del_license_hook() ->
+    _ = emqx_hooks:del('client.connect', {?MODULE, check, []}),
+    ok.
+
+do_update({file, Filename}, Conf) ->
+    case file:read_file(Filename) of
+        {ok, Content} ->
+            case emqx_license_parser:parse(Content) of
+                {ok, _License} ->
+                    Conf#{
+                      <<"type">> => <<"file">>,
+                      <<"value">> => Filename
+                     };
+                {error, Reason} ->
+                    error(Reason)
+            end;
+        {error, Reason} ->
+            error({invalid_license_file, Reason})
```

Comment:
Small nit: maybe the first is the `invalid_license_file` case, and the second one would be `license_not_found` ?